Sort upgraded CSS

RobinMalfait opened the pull request:
During the migration process, a lot of changes to the CSS file happen. Some parts are converted, some parts are deleted and some new CSS is added.

To make sure we are generating a sensible and good looking CSS file, we will sort the final CSS and pretty print it.

The order we came up with looks like this:

```css
/* Imports */
@import "tailwindcss";
@import "../other.css";

/* Configuration */
@config "../path/to/tailwindcss.config.js";

@plugin "my-plugin-1";
@plugin "my-plugin-2";

@source "./foo/**/*.ts";
@source "./bar/**/*.ts";

@variant foo {}
@variant bar {}

@theme {}

/* Border compatibility CSS */
@layer base {}

/* Utilities */
@utility foo {}
@utility bar {}

/* Rest of your own CSS if any */
```

RobinMalfait committed:
bucket sorting

RobinMalfait committed:
connect dangling comments to correct "bucket"

RobinMalfait committed:
update changelog

thecrypticace commented on packages/@tailwindcss-upgrade/src/codemods/sort-buckets.ts:
```diff
@@ -0,0 +1,142 @@
+        }
+
+        // Known at-rules
+        else if (node.type === 'atrule' && node.name === 'utility') {
+          injectInto('utility', node)
+        } else if (node.type === 'atrule' && node.name === 'variant') {
+          injectInto('variant', node)
+        } else if (node.type === 'atrule' && node.name === 'source') {
+          injectInto('source', node)
+        } else if (node.type === 'atrule' && node.name === 'theme') {
+          injectInto('theme', node)
+        } else if (node.type === 'atrule' && node.name === 'config') {
+          injectInto('config', node)
+        } else if (node.type === 'atrule' && node.name === 'plugin') {
+          injectInto('plugin', node)
+        }
```
Given that node.name always becomes the bucket name in these handful of cases I kinda feel like these could be simplified. I'm indifferent to actually doing this tho — just pointing it out.

thecrypticace commented on packages/@tailwindcss-upgrade/src/codemods/migrate-tailwind-directives.test.ts:
```diff
@@ -412,7 +414,7 @@ it('should replace `@responsive` with its children', async () => {
-      color: red;
-    }"
+        color: red;
+      }"
```
This is formatted kinda weird isn't it?

RobinMalfait replied:
Ah yep, I'll see if I can clean this up. What's happening is that:

1. This is part of the "user" CSS bucket, which we don't format
2. The `@responsive` just replaces the outer layer with its children. This results in CSS like this:

```css
··.format·{
····color:·red;
··}
```
Added `·` to show the spaces, but that's why it's being formatted this way.

RobinMalfait replied:
Solved it by marking each node inside `@responsive` with `tailwind_pretty`

RobinMalfait committed:
format user nodes in `@responsive`

Otherwise the wrong indentation is being used.

RobinMalfait committed:
use a loop instead of separate if-else-statements

RobinMalfait committed:
rename `@bucket` to `@tw-bucket`

Since this is purely an internal API, using a `@tw-` prefix reduces the
chances of colliding with real user CSS.

philipp-spiess commented on CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Ensure adjacent rules are merged together after handling nesting when generating optimized CSS ([#14873](https://github.com/tailwindlabs/tailwindcss/pull/14873))
 - _Upgrade (experimental)_: Install `@tailwindcss/postcss` next to `tailwindcss` ([#14830](https://github.com/tailwindlabs/tailwindcss/pull/14830))
 - _Upgrade (experimental)_: Remove whitespace around `,` separator when print arbitrary values ([#14838](https://github.com/tailwindlabs/tailwindcss/pull/14838))
+- _Upgrade (experimental)_: Sort upgraded CSS ([#14866](https://github.com/tailwindlabs/tailwindcss/pull/14866))
```
The current message makes it sound like the order was wrong or something but the change is only really for esthetic reasons, what do you think about this? I tried to not use the word order since it has an inherent meaning in CSS and am trying to lean on the "restructure" terminology

```suggestion
- _Upgrade (experimental)_: Restructure the generated CSS to make it easier to read ([#14866](https://github.com/tailwindlabs/tailwindcss/pull/14866))
```

Also this doesn't really fix anything so maybe it shouldn't be in the fixed category either I think 🤔 

I am still not convinced this is a very good message, maybe maybe this just doesn't need a changelog entry at all haha

RobinMalfait replied:
Yeah hmm, maybe this is just an internal refactor. Especially because the codemods are not idempotent right now meaning you only will be able to run them once anyway. So it's still a single 'upgrade' and not a fix to previous upgrades from the user's perspective.

philipp-spiess commented on integrations/upgrade/index.test.ts:
Reading this diff is bliss 😄

philipp-spiess commented on packages/@tailwindcss-upgrade/src/codemods/format-nodes.ts:
```diff
@@ -1,33 +1,73 @@
+
+      // Format buckets
+      if (node.type === 'atrule' && node.name === 'tw-bucket') {
+        output.push(await format(contents, { parser: 'css', semi: true, singleQuote: true }))
```
Maybe we can move out this part into a const object so we ensure we don't diverge the prettier config between the two branches (and avoid creating the same object over and over):

```suggestion
        output.push(await format(contents, FORMATTING_OPTIONS))
```

philipp-spiess commented on packages/@tailwindcss-upgrade/src/codemods/migrate-at-layer-utilities.ts:
```diff
@@ -192,7 +192,6 @@ export function migrateAtLayerUtilities(stylesheet: Stylesheet): Plugin {
       clone.params = cls
 
       // Mark the node as pretty so that it gets formatted by Prettier later.
```
```suggestion
```

philipp-spiess reviewed (commented):
Let's gooo it's so good to read the diff of the integration test snapshots haha. Some comments inline, mostly not sure about the change log thing 🤔

RobinMalfait committed:
re-use format options

RobinMalfait committed:
Update packages/@tailwindcss-upgrade/src/codemods/migrate-at-layer-utilities.ts

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
drop changelog entry
